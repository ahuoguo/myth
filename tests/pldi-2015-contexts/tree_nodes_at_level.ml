#use "context_bool.decls"

let tree_nodes_at_level : tree -> nat -> nat |>
  { Leaf =>
    ( 0 => 0
    | 1 => 0
    | 2 => 0
    | 3 => 0
    )
  | Node (Leaf, True, Leaf) =>
    ( 0 => 1
    | 1 => 0
    | 2 => 0
    | 3 => 0
    )
  | Node (Node (Leaf, True, Leaf), True, Leaf) =>
    ( 0 => 1
    | 1 => 1
    | 2 => 0
    | 3 => 0
    )
  | Node (Leaf, True, Node (Leaf, True, Leaf)) =>
    ( 0 => 1
    | 1 => 1
    | 2 => 0
    | 3 => 0
    )
  | Node (Node (Leaf, True, Leaf), True, Node (Leaf, True, Leaf)) =>
    ( 0 => 1
    | 1 => 2
    | 2 => 0
    | 3 => 0
    )
  | Node (Node (Node(Leaf, True, Leaf), True, Node (Leaf, True, Leaf)), True, Leaf) =>
    ( 0 => 1
    | 1 => 1
    | 2 => 2
    | 3 => 0)
  } = ?
