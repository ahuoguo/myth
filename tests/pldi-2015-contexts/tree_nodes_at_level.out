(* Generated Thu Nov 13 04:46:05 2014 *)
let tree_nodes_at_level : tree -> nat -> nat =
  let rec f1 (t1:tree) : nat -> nat =
    fun (n1:nat) ->
      match n1 with
        | O -> (match t1 with
                  | Leaf -> O
                  | Node (t2, b1, t3) -> S (O))
        | S (n2) -> (match t1 with
                       | Leaf -> O
                       | Node (t2, b1, t3) -> plus (f1 t2 n2) (f1 t3 n2))
  in
    f1
;;
